### PR TITLE
Slack Notifications - Bundle Publishing flows - Part 1

### PR DESCRIPTION
**Pull request by MaciekBaron**
### What is the context of this PR?

Loom Demo Video: https://www.loom.com/share/8ea9acc7e03f4faeaae636253e400e11

This PR related to CMS-366 and expands our Slack notification system. It covers the bundles ACs from the ticket, which are listed below.

Please not that the "failed to under pre-publish state" AC seems to be redundant, as discussed with Mebin, and therefore has not been addressed.

This change also includes adding two new bundle statuses, `FAILED` and `PARTIALLY_PUBLISHED`. These statuses behave similarly to `PUBLISHED`, being immutable end states for the bundle, but represent failures in the bundle pubish.
Commit [fe89d8e](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/fe89d8e9eb23ac63122e496a2825235e99b6f4e6) handles these new statuses in the admin UI, filtering them out from the default bundle listing and redirecting the edit URL (as we do for published bundles).
Commit [0a0f7f7](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/0a0f7f78539bbc6258b1d8efd48e62b349e8d7d3) skips syncing these new bundle statuses with the dataset bundle API.

### AC Checklist

- [x] Scheduled Bundle Pre-Publish Alert via Slack
- [ ] ~Scheduled Page Pre-Publish Alert via Slack~
- [x] Slack Alert: Bundle Publishing Commencement
- [ ] ~Slack Alert: Page Publishing Commencement~
- [x] Slack Alert: Bundle Publishing Finished Successfully
- [ ] ~Slack Alert: Page Publishing Finished Successfully~
- [ ] Bundle failed to enter “Pre-publish state" Alarm **(may be redundant?)**
- [x] Bundle Content Publication Failure (unexpected issues) Alarm
- [ ] ~Page Publication Failure  (unexpected issues) Alarm~
- [ ] ~Slack Notification for Technical Publishing Errors Alarm~ 
- [x] Slack Alert: Dataset API Call Failed
- [x] Third party API integration fail (generic) - currently used for Bundle API

### How to review

1. Ensure that all changes cover the bundle ACs listed in the ticket (with the exception noted above)
2. Create a Slack workspace and/or provide a working Slack bot API
3. Perform actions which trigger the messages
4. Confirm the messages are sent as expected

### Follow-up Actions

- [ ] Complete ACs related to pages
- [ ] Confirm that this functionality works with ONS Slack channels
- [ ] Configure the env vars after this PR merges

### CODE REVIEW

**Review by MebinAbraham — COMMENTED** [review]
Let's catch up, got some early thoughts on current approach.

**Review by AdamHawtin — COMMENTED** [review]
For posterity, a number of changes/fixes have been suggested in a tech session:
- Rename the channel settings variables for clarity to follow the pattern `SLACK_<CHANNEL_NAME>_CHANNEL`
- Move away from logging a warning every time a message send is attempted if the token/channel aren't configured. Check and warn on these at startup instead (which will both warn earlier and be less noisy in development envs with no slack integration). Perhaps this would be best done in a separate startup checks init container?
- If the channel/token are not configured, log the message and fields instead to allow for some level of local testing
- Switch off the pre-publication phase bundle status change notifications by default
- Ensure the first bundle publication status notification forces a new message, to make sure it pops up visibly in the channel

File: cms/bundles/management/commands/publish_bundles.py [hunk 1]
```diff
@@ -94,5 +95,6 @@ def handle(self, *args: Any, **options: Any) -> None:
                 bundle_scheduler.enterabs(bundle_ts, 1, self._handle_bundle_action, argument=(bundle,))
                 if bundle_ts > now_ts:
                     self.stdout.write(f"Publishing {bundle.name} in {bundle_ts - now_ts:.0f}s")
+                    notify_slack_of_bundle_pre_publish(bundle, bundle.release_date)
```
[the remark this answers is not in the document]
Comment (AdamHawtin) [reply cms/bundles/management/commands/publish_bundles.py]: I think that is the expected behaviour in this case, the pre-publication phase is expected to happen slightly before publish, then the publish start messages will signal the beginning of the actual bundle publish.

**Review by BJacksonONS — APPROVED** [review]
Tested with my own slack workspace, all working as expected.

**Review by MebinAbraham — COMMENTED** [review]
Could you add a doc to `docs/integrations/` folder outlining our usages and set-up.

**Comment by MaciekBaron** [comment]
@MebinAbraham updated docs here (not sure how much detail to include): [f47162e](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/f47162eb55d062c942a1383b8f55b043fea72da0)

File: cms/core/tests/test_slack.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    @override_settings(SLACK_BOT_TOKEN="xoxb-test-token")
+    @patch("cms.core.slack.WebClient")
+    def test_get_slack_client__with_token(self, mock_client):
+        """Should return WebClient instance when token is configured."""
+        client = get_slack_client()
+        self.assertIsNotNone(client)
+        mock_client.assert_called_once_with(token="xoxb-test-token")
+
+    @override_settings(SLACK_BOT_TOKEN=None)
+    def test_get_slack_client__without_token(self):
```
Comment (sanjeevz3009) [inline cms/core/tests/test_slack.py:184]: What's the difference between this test and the test below? They seem to be doing the same thing?

**Review by sanjeevz3009 — COMMENTED** [review]
Did a round of code review. So far, it looks good. Will do another round and functional test review tomorrow morning before approving.

@MaciekBaron @MebinAbraham

File: cms/core/tests/test_slack.py (hunk 2 above)
Comment (MaciekBaron) [reply cms/core/tests/test_slack.py:184]: Not sure, looks like Adam added it by accident, removed the second one.

[840c45c](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/840c45c93c73448656d030b49ce2a5e2a8b4fcc7)

File: cms/bundles/clients/api.py [hunk 3]
```diff
@@ -142,14 +148,42 @@ def _make_request(  # pylint: disable=too-many-arguments  # noqa: PLR0913
                 extra=logger_extra,
             )
 
+            # Determine alert type based on status code
+            alert_type = (
+                BundleAlertType.CRITICAL
+                if e.response.status_code >= HTTPStatus.INTERNAL_SERVER_ERROR
```
[the remark this answers is not in the document]
Comment (MaciekBaron) [reply cms/bundles/clients/api.py]: Happy to change it to 400: [fdda0e1](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/fdda0e1301a4bdcf085a6350509f15d4aaac8ffb)

File: cms/core/slack.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    Returns:
+        WebClient instance if SLACK_BOT_TOKEN is configured, None otherwise.
+    """
+    if not (token := settings.SLACK_BOT_TOKEN):
+        return None
+    return WebClient(token=token)
+
+
+def send_or_update_message(
```
Comment (MebinAbraham) [inline cms/core/slack.py]: Nit: this would make sense as a method of a class, or if used as `slack.send_or_update_message`, otherwise, it is hard to know what this is referring to. make update to `send_or_update_slack_message`

Comment (MaciekBaron) [reply cms/core/slack.py]: Renamed: [ceee535](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/ceee53555da3983bf92b9ee10c7b391991469011)

File: cms/bundles/utils.py [hunk 5]
```diff
@@ -436,6 +448,7 @@ def publish_bundle(bundle: Bundle, *, update_status: bool = True) -> bool:
                         exception_message=f"<{page.full_edit_url}|Page (ID: {page.pk})> in the bundle did not "
                         "publish because it is not in a workflow or has no revisions",
                     )
+                    continue
```
Comment (BJacksonONS) [inline cms/bundles/utils.py]: Nit: If pages_published is incremented in the if/elif branches we don't need the continue here

File: cms/bundles/utils.py [hunk 6]
```diff
@@ -431,65 +443,96 @@ def publish_bundle(bundle: Bundle, *, update_status: bool = True) -> bool:
...
+            )
+            alert_slack_of_bundle_content_failure(
+                bundle=bundle,
+                exception_message=f"<{page.full_edit_url}|Page (ID: {page.pk})> in the bundle failed to publish",
             )
-            failed_page_publishes.append(page.pk)
 
     # update and publish related release calendar
     if bundle.release_calendar_page_id:
         update_bundle_linked_release_calendar_page(bundle)
```
Comment (BJacksonONS) [inline cms/bundles/utils.py:469]: I'm not sure if this is actually different from previous logic, but we're updating the release calendar page and publishing before we check if there was a failure publishing the bundle.

Is that what we want?

File: cms/bundles/utils.py (hunk 5 above)
Comment (MaciekBaron) [reply cms/bundles/utils.py]: [486f51f](https://github.com/ONSdigital/dis-wagtail/pull/504/commits/486f51f47c74db18a306c2d2f9caa01f93c73f33)

File: cms/bundles/utils.py (hunk 6 above)
Comment (MebinAbraham) [reply cms/bundles/utils.py:469]: Yes, we want to be optimistic. Some pages may have been published, however, even if all pages failed, they may contain a dataset which we don't control. So, in theory, a scenario where we may not want to publish the RC is when the `page_published` count is 0 and the bundle has no datasets. But we would need an analysis of what the business expects to happen, so let's revisit. Can you make a small card to flag this?

File: cms/core/models/base.py [hunk 7]
```diff
@@ -309,6 +310,15 @@ def _get_site_root_paths(self, request: HttpRequest | models.Model | None = None
...
             return paths
 
+    @property
+    def full_edit_url(self) -> str | None:
+        """Returns the absolute URL for the page edit view, or an empty string if the page is not saved yet."""
+        if not self.pk:
+            return ""
+
+        base_url = settings.WAGTAILADMIN_BASE_URL
+        return f"{base_url}{reverse('wagtailadmin_pages:edit', args=[self.pk])}"
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/core/models/base.py]: Resolving on behalf!

File: cms/bundles/clients/api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/bundles/clients/api.py]: Resolving on behalf!

File: cms/bundles/models.py [hunk 8]
```diff
@@ -156,6 +156,7 @@ class Bundle(index.Indexed, ClusterableModel, models.Model):  # type: ignore[dja
     # note: it looks like etag is SHA-1, so 40 chars, but using 255 for safety
     # https://github.com/ONSdigital/dp-net/blob/a17216881f99417aefa7aa256a337e2ad635866d/handlers/response/etag.go#L13
     bundle_api_etag = models.CharField(max_length=255, blank=True, editable=False)
+    slack_notification_ts = models.CharField(max_length=32, blank=True, null=True, editable=False)
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/bundles/models.py]: Resolving on behalf!

File: cms/bundles/notifications/slack.py [hunk 9]
```diff
@@ -1,114 +1,348 @@
...
+        bundle: The bundle to extract context from.
+
+    Returns:
+        Dictionary containing publish_type, page_count, and example_page_url.
+    """
+    return {
+        "publish_type": _get_publish_type(bundle),
+        "page_count": bundle.get_bundled_pages().count(),
+        "example_page_url": _get_example_page_url(bundle),
+    }
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/bundles/notifications/slack.py]: Resolving!

File: cms/bundles/notifications/slack.py [hunk 10]
```diff
@@ -1,114 +1,348 @@
...
+        fields.append({"title": "Example Page", "value": context["example_page_url"], "short": False})
+
+    text = "Publishing the bundle has started"
+
+    send_bundle_notification(
+        bundle=bundle,
+        text=text,
+        color="warning",  # Amber
+        fields=fields,
+    )
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/bundles/notifications/slack.py]: Resolving!

File: cms/bundles/utils.py [hunk 11]
```diff
@@ -378,10 +378,19 @@ def publish_bundle(bundle: Bundle, *, update_status: bool = True) -> bool:
...
     notifications = __import__(
-        "cms.bundles.notifications.slack", fromlist=["notify_slack_of_publication_start", "notify_slack_of_publish_end"]
+        "cms.bundles.notifications.slack",
+        fromlist=[
+            "BundleAlertType",
+            "notify_slack_of_publication_start",
+            "notify_slack_of_publish_end",
+            "notify_slack_of_bundle_failure",
+        ],
     )
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/bundles/utils.py]: Resolving!

File: cms/bundles/utils.py [hunk 12]
```diff
@@ -391,22 +400,34 @@ def publish_bundle(bundle: Bundle, *, update_status: bool = True) -> bool:
...
+        start_time=bundle.scheduled_publication_date or start_time,
+        url=bundle.full_inspect_url,
+    )
+    # Track successful and failed page publications
+    pages_published = 0
+    failed_pages: list[int] = []
+    total_pages = 0
 
     for page in bundle.get_bundled_pages().specific(defer=True).select_related("latest_revision").not_live():
+        total_pages += 1
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/bundles/utils.py]: Resolving!

File: cms/core/slack.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        else:
+            # Create new message
+            response = client.chat_postMessage(
+                channel=channel,
+                text=text,
+                attachments=attachments,
+                unfurl_links=False,
+                unfurl_media=False,
+            )
+            # Return the
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/core/slack.py]: Resolving!

File: cms/datasets/models.py [hunk 14]
```diff
@@ -70,28 +76,64 @@ def fetch_api_response(self, url: str | None = None, params: Mapping | None = No
         except requests.exceptions.HTTPError as e:
             if e.response.status_code == HTTPStatus.TOO_MANY_REQUESTS:
                 logger.warning("Rate limit exceeded when fetching datasets", extra={"url": url})
+                notify_slack_of_dataset_api_failure(
+                    page=None,
+                    exception_message=f"Rate limit exceeded: HTTP {e.response.status_code}",
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/datasets/models.py]: Resolving!

File: cms/datasets/models.py [hunk 15]
```diff
@@ -70,28 +76,64 @@ def fetch_api_response(self, url: str | None = None, params: Mapping | None = No
...
+                alert_type=BundleAlertType.WARNING,
+            )
             raise ValueError("Failed to parse JSON response from datasets API") from e
 
         # api_response should be a dict, let's raise a clear error if not
         if not isinstance(api_response, dict):
             logger.error("Invalid API response format when fetching datasets", extra={"url": url, "params": params})
+            notify_slack_of_dataset_api_failure(
+                page=None,
+                exception_message="Invalid API response format, expected a dictionary-like object",
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/datasets/models.py:134]: Resolving!

File: cms/core/slack.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        if update_message_ts:
+            try:
+                response = client.chat_postMessage(
+                    channel=channel,
+                    text=text,
+                    attachments=attachments,
+                    unfurl_links=False,
+                    unfurl_media=False,
+                )
+                if response and response.get("ok") and response.get("ts"):
```
[the remark this answers is not in the document]
Comment (sanjeevz3009) [reply cms/core/slack.py]: Resolving!

File: cms/core/tests/test_slack.py (hunk 2 above)
Comment (sanjeevz3009) [reply cms/core/tests/test_slack.py:184]: Thanks @MaciekBaron

File: cms/bundles/utils.py (hunk 6 above)
Comment (sanjeevz3009) [reply cms/bundles/utils.py:469]: @BJacksonONS Did you end up creating a card? If so, can you chuck it in this thread, for posterity please =)

Comment (BJacksonONS) [reply cms/bundles/utils.py:469]: https://officefornationalstatistics.atlassian.net/browse/CMS-1235

File: cms/bundles/tests/test_slack_notifications.py [hunk 17]
```diff
@@ -1,170 +1,753 @@
...
     notify_slack_of_status_change,
+    send_bundle_notification,
 )
 from cms.bundles.tests.factories import BundleFactory
+from cms.release_calendar.tests.factories import ReleaseCalendarPageFactory
 from cms.users.tests.factories import UserFactory
 
 
-class SlackNotificationsTestCase(TestCase):
+class SendBundleNotificationTestCase(TestCase):
```
Comment (sanjeevz3009) [inline cms/bundles/tests/test_slack_notifications.py:33]: Since you did it below for the `BundleStatusNotificationsTestCase`. Add a doc string for this test class as well, please.

**Review by sanjeevz3009 — COMMENTED** [review]
Final round of code review complete. A minor comment, around having doc strings for test classes, not all of them need them, but I pointed out one.

I will functionally test this using the test cases we have already, and some other edge cases with @BJacksonONS tomorrow morning, before approving.

File: cms/settings/base.py [hunk 18]
```diff
@@ -977,6 +977,13 @@
 START_TIME = datetime.datetime.now(tz=datetime.UTC)
 
 SLACK_NOTIFICATIONS_WEBHOOK_URL = env.get("SLACK_NOTIFICATIONS_WEBHOOK_URL")
```
Comment (sanjeevz3009) [inline cms/settings/base.py:979]: We don't seem to be utilising this truly anywhere?

Is it not set in the helm charts either? Do we still need it?

I see it in some test files

@MaciekBaron

File: cms/bundles/notifications/slack.py [hunk 19]
```diff
@@ -1,114 +1,312 @@
...
+    Returns the release calendar page URL if available,
+    otherwise the first bundled page URL, excluding alias pages.
+
+    Args:
+        bundle: The bundle to get the example page URL for.
+
+    Returns:
+        The example page URL or None if no pages available.
+    """
+    if release_page := bundle.release_calendar_page:
```
Comment (BJacksonONS) [inline cms/bundles/notifications/slack.py:103]: I don't know if it's a show stopper but noticed here that `full_url` is getting the first site host, rather than the default, so example page urls are often using the welsh subdomain.

Comment (BJacksonONS) [reply cms/bundles/notifications/slack.py:103]: @sanjeevz3009 thought this might have been raised before but I can't see any comments/commits around it.

Comment (sanjeevz3009) [reply cms/bundles/notifications/slack.py:103]: @MaciekBaron, I think Adam had a conversation with you about this issue, where welsh sub domain url gets added to the Slack notification for all types of bundle publication. The issue seemed to have been fixed with the commit https://github.com/ONSdigital/dis-wagtail/pull/504/changes/9d56af404b3d216f32c9f929e61c5c69abac5b23

I can't reproduce this locally as it works fine for me, but when Ben publishes bundles, the Slack notifications contain the Welsh subdomain URLs.

Comment (BJacksonONS) [reply cms/bundles/notifications/slack.py:103]: Looks like its something around the default site setup from the data migrations and the default ordering applied when using `get_site_root_paths`.

image

File: cms/bundles/clients/api.py [hunk 20]
```diff
@@ -142,14 +148,42 @@ def _make_request(  # pylint: disable=too-many-arguments  # noqa: PLR0913
...
             raise BundleAPIClientError(error_msg, errors) from e
 
         except requests.exceptions.RequestException as e:
+            # Import here to avoid circular import
+            from cms.bundles.notifications.api_failures import (  # pylint: disable=import-outside-toplevel
+                notify_slack_of_third_party_api_failure,
+            )
+            from cms.bundles.notifications.slack import BundleAlertType  # pylint: disable=import-outside-toplevel
+
             error_msg = f"Network error for {method} {url}: {e!s}"
```
Comment (kevinhowbrook) [inline cms/bundles/clients/api.py]: **Issue?**

Should this use `_format_http_error`

Lower down, we specifically strip the URL in self._format_http_error (L236):

```python
# This message is shown to the user, we don't want to leak URLs or sensitive info
        formatted_msg = f"HTTP {status_code} error: {status_phrase}"
```

Doesn't this notification leak the URL here?

**Review by sanjeevz3009 — APPROVED** [review]
Wrapped my testing locally. I used different test scripts to simulate the behaviours. I can see the Slack alerts on the alarm and log channels for the narrowed scope of this PR/ ticket.

We need to make it clear on the `CMS-366` ticket the true scope of the work being done here as part of it, as some of the ACs have been extracted to `Slack 2, 3, and 4` follow-up cards.

Other than that, LGTM, apart from the minor comments. The main thing about turning this on is that if there were any issues with Slack notification firing failing itself, then bundle publication won't be affected, which is excellent and what we want. I tested this too via a test script.

**Comment by sanjeevz3009** [comment]
@BJacksonONS 

In the absence of @MaciekBaron let's clean this PR's final comments up between you, me and @kevinhowbrook and get this merged tomorrow morning =)

File: cms/bundles/tests/test_slack_notifications.py (hunk 17 above)
Comment (sanjeevz3009) [reply cms/bundles/tests/test_slack_notifications.py:33]: I have added the minor comments as follow-ups to be done as part of the next Slack tickets.

File: cms/settings/base.py (hunk 18 above)
Comment (sanjeevz3009) [reply cms/settings/base.py:979]: I have added the minor comments as follow-ups to be done as part of the next Slack tickets.

File: cms/bundles/notifications/slack.py (hunk 19 above)
Comment (sanjeevz3009) [reply cms/bundles/notifications/slack.py:103]: Will see how this behaves in the deployed environment.

File: cms/bundles/clients/api.py (hunk 20 above)
Comment (sanjeevz3009) [reply cms/bundles/clients/api.py]: @BJacksonONS and I looked into this and went with this minimal change approach: https://github.com/ONSdigital/dis-wagtail/pull/504/commits/5174b35b82cecc9cfcbcaa783f23c2f829e23041

As the `_format_http_error` private helper method is for HTTP errors, not network errors, etc., we didn't want to make changes to that method.

We went with keeping logging the URL for our own logs, open search, etc., and removing the URL from the Slack messages.

**Review by BJacksonONS — APPROVED** [review]
LGTM

**Review by kevinhowbrook — APPROVED** [review]

**Review by kevinhowbrook — APPROVED** [review]
